hudi-common/src/main/java/org/apache/hudi/io/storage/HoodieNativeAvroHFileReader.java

```diff
@@ -244,14 +248,32 @@ private static GenericRecord getRecordFromKeyValue(KeyValue keyValue,
         readerSchema);
   }
 
-  private synchronized HFileReader getSharedHFileReader() {
-    try {
-      if (!sharedHFileReader.isPresent()) {
-        sharedHFileReader = Option.of(newHFileReader());
+  private byte[] getHFileMetaInfoFromCache(String key) throws IOException {
+    if (!PRELOADED_META_INFO_KEYS.contains(key)) {
+      throw new IllegalStateException("HoodieNativeAvroHFileReader#getHFileMetaInfoFromCache"
+          + " should only be called on supported meta info keys; this key is not supported: "
+          + key);
+    }
+    byte[] bytes = metaInfoMap.get(key);
+    if (bytes != null) {
+      return bytes;
+    }
+    loadAllMetaInfoIntoCache();
+    return metaInfoMap.get(key);
+  }
+
+  private synchronized void loadAllMetaInfoIntoCache() throws IOException {
+    // Load all meta info that are small into cache
```

Comment:
I think there also might need to be a check here to see if the cache is loaded and to skip reloading

Reply:
Addressed.